cpp/include/raft/sparse/neighbors/detail/connect_components.cuh

```diff
@@ -172,46 +179,80 @@ struct LookupColorOp {
  * the given array of components
  * @tparam value_idx
  * @tparam value_t
+ * @param[in] handle raft handle
  * @param[out] kvp mapping of closest neighbor vertex and distance for each vertex in the given
  * array of components
  * @param[out] nn_colors components of nearest neighbors for each vertex
  * @param[in] colors components of each vertex
  * @param[in] X original dense data
  * @param[in] n_rows number of rows in original dense data
  * @param[in] n_cols number of columns in original dense data
- * @param[in] stream cuda stream for which to order cuda operations
+ * @param[in] batch_size batch size for computing nearest neighbors
+ * @param[in] reduction_op reduction operation for computing nearest neighbors
  */
 template <typename value_idx, typename value_t, typename red_op>
-void perform_1nn(raft::KeyValuePair<value_idx, value_t>* kvp,
+void perform_1nn(raft::device_resources const& handle,
+                 raft::KeyValuePair<value_idx, value_t>* kvp,
                  value_idx* nn_colors,
                  value_idx* colors,
                  const value_t* X,
                  size_t n_rows,
                  size_t n_cols,
-                 cudaStream_t stream,
+                 size_t batch_size,
                  red_op reduction_op)
 {
-  rmm::device_uvector<int> workspace(n_rows, stream);
-  rmm::device_uvector<value_t> x_norm(n_rows, stream);
-
-  raft::linalg::rowNorm(x_norm.data(), X, n_cols, n_rows, raft::linalg::L2Norm, true, stream);
-
-  raft::distance::fusedL2NN<value_t, raft::KeyValuePair<value_idx, value_t>, value_idx>(
-    kvp,
-    X,
-    X,
-    x_norm.data(),
-    x_norm.data(),
-    n_rows,
-    n_rows,
-    n_cols,
-    workspace.data(),
-    reduction_op,
-    reduction_op,
-    true,
-    true,
-    stream);
+  auto stream = handle.get_stream();
 
+  value_idx n_components = get_n_components(colors, n_rows, stream);
+  auto colors_group_idxs = raft::make_device_vector<value_idx, value_idx>(handle, n_components + 1);
+  raft::sparse::convert::sorted_coo_to_csr(
+    colors, n_rows, colors_group_idxs.data_handle(), n_components + 1, stream);
+
+  auto kvp_view =
+    raft::make_device_vector_view<raft::KeyValuePair<value_idx, value_t>, value_idx>(kvp, n_rows);
+  using OutT   = raft::KeyValuePair<value_idx, value_t>;
+  using ParamT = raft::distance::masked_l2_nn_params<red_op, red_op>;
+
+  ParamT params{reduction_op, reduction_op, true, true};
+
+  size_t n_batches = raft::ceildiv(n_rows, batch_size);
+  for (size_t bid = 0; bid < n_batches; bid++) {
+    size_t batch_offset   = bid * batch_size;
+    size_t rows_per_batch = min(batch_size, n_rows - batch_offset);
+    auto x_norm           = raft::make_device_vector<value_t, value_idx>(handle, rows_per_batch);
```

Comment:
It's better to allocate the x_norm vector once outside of the loop and also calculate the row norm once. Many small allocations and short kernels are inefficient. The memory usage will also not increase too much from allocating all of x_norm in one go.